### DISCUSSION TITLE
Renovate is not displaying changelog for helm charts using OCI, while it's working with helm repository

### DISCUSSION THREAD

**Discussion by n0rad** [renovatebot/renovate#25772] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

When you declare a dependency in an helm chart using OCI repository, instead of helm repository, renovate is able to propose update but do not display the changelog.

I created an example to reproduce the issue [here](https://github.com/n0rad/renovate-test/pull/2)

The [chart.yaml](https://github.com/n0rad/renovate-test/commit/74fa31312e0726ce39672fe252021471087e1729) is declared using a dependency from an helm repository. The PR is displaying the [changelog](https://github.com/n0rad/renovate-test/pull/2) (in the history of the PR description).

Then I switched the registry in the [charts.yaml](https://github.com/n0rad/renovate-test/blob/main/chart/Chart.yaml) to an [oci image](https://registry.hub.docker.com/r/n0rad/argo-cd/tags). This chart was fetch from the helm repository, and pushed to docker hub using helm 3.12.0, so the oci image contains opencontainers annotations.

it look like renovate is able to find the `org.opencontainers.image.url` and tries to fetch the changelog looking at the [logs](https://developer.mend.io/github/n0rad/renovate-test/-/job/ffca3c36-7975-43b1-b754-72b2d5125e63):
```
DEBUG: Fetching changelog: https://github.com/argoproj/argo-helm (5.50.0 -> 5.51.0) (branch="renovate/argo-cd-5.x")
```

but nothing is displayed in the PR.

**Comment by rarkins** [comment]
Looks like the releases are prefixed and renovate doesn't match them. Not helm specific if so

**Reply by n0rad** [reply]
I was actually doing tests to figure out if renovate is able to understand the release prefix in case of multitenant repositories, while i discovered this issue.
Note that with an helm repository, it's finding the release, even if it's prefixed.

**Reply by usmonster** [reply]
💁 Regarding this topic, I've opened #29943.

**Comment by n0rad** [comment]
@rarkins could you review your assumption? because to me the problem looks small and unrelated to the prefix.

It's working fine when using an `helm repository` (even if there is a prefix). 
It's finding the repository and trigger the changelog fetch when using a `OCI repository`, but do not get into displaying it.

**Reply by rarkins** [reply]
Unfortunately I don't have any time to look into this topic, but if you can find a simple fix for it then you're most welcome to submit a PR and we'll review that

**Reply by usmonster** [reply]
Looked into this and found some working counterexamples in the wild of Helm chart OCI updates with changelogs in the Renovate PRs, [collected a few here](https://github.com/usmonster/renovate-test/blob/main/chart/Chart.yaml) (and [related PRs](https://github.com/usmonster/renovate-test/pulls?q=is%3Aopen+is%3Apr)).

Got curious about why the example wasn't working, dug a bit, and—tl;dr—the OCI repository for the non-working example here is missing an intermediate version of the upstream chart (i.e. [missing 5.50.1 in this repo](https://hub.docker.com/r/n0rad/argo-cd/tags), but OK when using [this repo](https://hub.docker.com/repository/docker/usmonster/argo-cd/general)). So Renovate seems to abandon building the changelog if one of the versions found at the Chart `sources[0]` isn't present in the OCI `repository`, which makes sense.

Maybe some additional debug logs when changelog generation is abandoned could help in this case? Would a PR to add such logs be welcome, @rarkins? (I assume somewhere in [`getChangeLogJSON()`](https://github.com/renovatebot/renovate/blob/7516af6e6aface68ca3b1d605162ff4e9af2b928/lib/workers/repository/update/pr/changelog/source.ts#L67), but any tips would be much appreciated. 😄)

This would be especially helpful since our own current issue more concerns missing changelogs in PRs for updates to OCI Helm Chart dependencies our _private repos_ (source+tags+releases on GitHub, OCI images on Google Artifact Registry), and even after checking over [a dozen possibilities](https://docs.renovatebot.com/key-concepts/changelogs/#troubleshooting-missing-changelogs), we can't pinpoint why this isn't working in our setup.

Thanks!

**Reply by viceice** [reply]
I think we should use the platform for changelog source url to retrieve versions (eg github-tags) instead of using the source datasource (eg docker).

@rarkins @secustor WDYT?

**Reply by secustor** [reply]
Wouldn't drop supporting having Changelog files on generic endpoints?  Something like `https://myexample.com/foo/bar/Changelog.md` which then return the raw file as text.

**Reply by viceice** [reply]
no, it shouldn't. we are already in platform dependent code, so we know we're on GitHub and know the repo, but then we call datasource getReleases with original dependency to fetch all versions instead directly use github-tags. we use github-tags later to fetch release notes with  only known versions

**Reply by usmonster** [reply]
This sounds like a good solution (though still not sure why it works for other OCI Helm charts); would a PR be welcome for this / can an issue be generated?

In the meantime, do you see a workaround to force the datasource used in this case? I've thought about using a [custom datasource](https://docs.renovatebot.com/configuration-options/#customdatasources:~:text=overwrite%20default%20datasources), but I suspect it might break other things and have an effect not limited to the changelog generation part.

Thanks again!

**Reply by viceice** [reply]
- https://github.com/renovatebot/renovate/discussions/29687
